### ISSUE TITLE
etcd2 discovery + node failure + proxy promotion

### ISSUE THREAD

**Issue by JeanMertz** [coreos/etcd#2807]
I simulated the following situation:

* create 3-node etcd2 cluster
  ```
  % cluster is healthy
  member xxx is healthy
  member yyy is healthy
  member zzz is healthy
  ```

* destroy 1 node
  * one member now shows as `failed`
* create new node
  * node detects full cluster, becomes a proxy
  * after a while, cluster shows two members

    ```
    % cluster is healthy
    member xxx is healthy
    member yyy is healthy
    ```
 
If I restart etcd2 at this point, I get:

```
% systemctl restart etcd2
systemd[1]: Stopping etcd2...
etcd2[2475]: 2015/05/10 22:31:22 received terminated signal, shutting down...
systemd[1]: Started etcd2.
systemd[1]: Starting etcd2...
etcd2[2497]: 2015/05/10 22:31:22 etcd: already initialized as proxy before, starting as etcd proxy...
```

If I remove `/var/lib/etcd2/proxy`, I get:

```
% systemctl restart etcd2
systemd[1]: Stopping etcd2...
etcd2[2475]: 2015/05/10 22:32:15 received terminated signal, shutting down...
systemd[1]: Started etcd2.
systemd[1]: Starting etcd2...
etcd2[2522]: 2015/05/10 22:32:18 etcd: listening for peers on http://10.13.1.89:2380
etcd2[2522]: 2015/05/10 22:32:18 etcd: listening for client requests on http://0.0.0.0:2379
etcd2[2522]: 2015/05/10 22:32:18 etcdserver: datadir is valid for the 2.0.1 format
etcd2[2522]: 2015/05/10 22:32:19 etcd: stopping listening for client requests on http://0.0.0.0:2379
etcd2[2522]: 2015/05/10 22:32:19 etcd: stopping listening for peers on http://10.13.1.89:2380
etcd2[2522]: 2015/05/10 22:32:19 etcd: discovery cluster full, falling back to proxy
```

My question is, when using a discovery token, how can I promote this new node to become part of the cluster instead of a proxy? It seems like the discovery method can't be used in this case?

**Comment by xiang90** [comment]
@JeanMertz You are right. Discovery service can only be used for the first bootstrap. When you already bootstrapped your cluster, you should be able to manage it yourselves by listing the members. 

See https://github.com/coreos/etcd/blob/master/Documentation/clustering.md#lifetime-of-a-discovery-url

**Comment by JeanMertz** [comment]
@xiang90 yes, I read that chapter, and I also read about using `etcd2ctl member` to manage memberships, but I am unsure how to tie the two with the usage of CoreOS' user-data bootstrapping.

For example, I have this user-data for all etcd2 cluster machines:

```
coreos:
  etcd2:
    discovery: https://discovery.etcd.io/TOKEN
    advertise-client-urls: http://$private_ipv4:2379
    initial-advertise-peer-urls: http://$private_ipv4:2380
    listen-client-urls: http://0.0.0.0:2379
    listen-peer-urls: http://$private_ipv4:2380
  units:
  - name: etcd2.service
    command: start
```

but can you explain to me, once I completed the `create new node` step as described above, using the above user-data. How do I then get this pre-configured new etcd2 node to join as a member of the cluster, instead of a proxy?

Basically, using all the best-practice guides (of which most are from the CoreOS team itself), how do I merge these guides with the documentation described here:

https://github.com/coreos/etcd/blob/master/Documentation/runtime-configuration.md#add-a-new-member

without having to use separate user-data files to differentiate between "fresh" etcd2 cluster machines, and machines booted after an existing cluster has started. I don't care if it requires manual steps, I just like to avoid having to change the user-data to boot the machine.

**Comment by JeanMertz** [comment]
I created a `etcd2-rejoin.service` unit file, executing the following shell script:

```
[Service]
User=etcd
Environment=ETCD_DATA_DIR=/var/lib/etcd2
Environment=ETCD_NAME=%m
Environment=ETCD_INITIAL_CLUSTER_STATE=existing
Restart=always
RestartSec=10s
LimitNOFILE=40000

ExecStart=/usr/bin/bash -c "                                          \
  export ETCD_ADVERTISE_CLIENT_URLS=http://$private_ipv4:2379;        \
  export ETCD_INITIAL_ADVERTISE_PEER_URLS=http://$private_ipv4:2380;  \
  export ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379;                 \
  export ETCD_LISTEN_PEER_URLS=http://$private_ipv4:2380;             \
  systemctl start etcd2 && sleep 2;                                   \
                                                                      \
  details=$(etcdctl member add $ETCD_NAME http://$private_ipv4:2380); \
  eval export `echo "$details" | tail -n-3`;                          \
                                                                      \
  systemctl stop etcd2;                                               \
  rm -rf /var/lib/etcd2/proxy;                                        \
  /usr/bin/etcd2
```

This feels cumbersome, but going by the documentation, this seems the only way to semi-automate this. With this, we still call `systemctl start etcd2-rejoin`, which requires an explicit decision (which as far as I can tell was the main reason not to automate rejoining anymore), but at least the steps to take are reduced to a single command.

Any thoughts on this?

**Comment by xiang90** [comment]
@JeanMertz 

The short answer is:
if you want to use the same configuration to bootstrap, you'd better to know the exact number of members you want. 

Dynamic configuration change is only used for replacing dead members in most cases.  

So when you bootstrap, you can:
1. bootstrap members and proxies at the same time using discovery

2. bootstrap members first using static configuration if you can then bootstrap proxies.

The benefits of 2 is you can have a separate etcd cluster. We suggest 2 if you have a large deployment. 

We also see people write their own etcd cluster management tools to deal with dynamic configuration automation. But we have no short term plan to make it happen.
